cmake/modules/root.cmake

```diff
@@ -32,3 +32,10 @@ zephyr_file(APPLICATION_ROOT SOC_ROOT)
 
 # Convert paths to absolute, relative from APPLICATION_SOURCE_DIR
 zephyr_file(APPLICATION_ROOT ARCH_ROOT)
+
+if(unittest IN_LIST Zephyr_FIND_COMPONENTS)
+  # Zephyr used in unittest mode, use dedicated unittest root.
+  set(BOARD_ROOT ${ZEPHYR_BASE}/tests/root)
+  set(ARCH_ROOT  ${ZEPHYR_BASE}/tests/root)
+  set(SOC_ROOT   ${ZEPHYR_BASE}/tests/root)
+endif()
```

Comment:
Maybe set these in unittest.cmake?

Reply:
no, the introduction of Zephyr CMake modules and its modularization here https://github.com/zephyrproject-rtos/zephyr/pull/41301 tries to improve the overall design and reuse of code.

This file deals with root handling, that is used by other modules.
If you set those in unittest.cmake you must set those BEFORE any `include()`, but that breaks the design as you loose some features of the modular design.
All Zephyr CMake modules (except unittest and zephyr_default that a special cases) are having include guards:
https://github.com/zephyrproject-rtos/zephyr/blob/ccc0315e14341879cc466156a4c925dd27b06525/cmake/modules/root.cmake#L20

This allows code to include them multiple times and only have them executed the first time.
This again means that users can do neat trick like this, before loading, for example unittest.

Normal:
```
find_package(Zephyr COMPONENTS unittest)
```

Locally adjusted root result:
```
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/modules_tweaks)
find_package(Zephyr COMPONENTS root root_tweak unittest)
```

and then inside `root_tweak` decide what should happen to any root.
Then when `unittest` is loaded, the result from `root_tweak` is used.
Such possibility is lost if we go back to old style boilerplate of placing everything in just one file.

The cleanup and design work done in #41301 is exactly what make it possible to easily reuse Kconfig know.
Although there are still a lot of improvement to be done.

For a similar example, look here:
https://github.com/zephyrproject-rtos/zephyr/blob/ccc0315e14341879cc466156a4c925dd27b06525/share/sysbuild/CMakeLists.txt#L16-L23

Reply:
That's really neat, thanks for the detailed explanation, looks like I have some more to learn about our build system :)